ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/bridge/ConflictCheckingLedgerBridge.scala

```diff
@@ -0,0 +1,362 @@
+          }
+    }
+
+  private def validatePartyAllocation(
+      transaction: Submission.Transaction
+  )(implicit
+      contextualizedErrorLogger: ContextualizedErrorLogger
+  ): AsyncValidation[Unit] = {
+    val _ = (transaction, contextualizedErrorLogger)
+    // TODO SoX: Implement
+    Future.successful(Right(()))
+  }
+
+  private def checkTimeModel(
+      transaction: Transaction
+  ): Validated[Unit] = {
+    // TODO SoX: Implement
+    val _ = transaction
+    Right(())
+  }
+
+  private def validateCausalMonotonicity(
+      transaction: Transaction,
+      inputContracts: Set[ContractId],
+      transactionLedgerEffectiveTime: Timestamp,
+      divulged: Set[ContractId],
+  )(implicit
+      contextualizedErrorLogger: ContextualizedErrorLogger
+  ): AsyncValidation[Unit] = {
+    val referredContracts = inputContracts.diff(divulged)
+    if (referredContracts.isEmpty)
+      Future.successful(Right(()))
+    else
+      indexService
+        .lookupMaximumLedgerTime(referredContracts)
+        .transform {
+          case Failure(MissingContracts(missingContractIds)) =>
+            Success(
+              Left(
+                toCommandRejectedUpdate(
+                  UnknownContracts(missingContractIds)(errorFactories),
+                  transaction.submitterInfo.toCompletionInfo,
+                )
+              )
+            )
+          case Failure(err) =>
+            Success(
+              Left(
+                toCommandRejectedUpdate(
+                  LedgerBridgeInternalError(err),
+                  transaction.submitterInfo.toCompletionInfo,
+                )
+              )
+            )
+          case Success(maximumLedgerEffectiveTime) =>
+            maximumLedgerEffectiveTime
+              .filter(_ > transactionLedgerEffectiveTime)
+              .fold[Try[Validated[Unit]]](Success(Right(())))(contractLedgerEffectiveTime =>
+                Success(
+                  Left(
+                    toCommandRejectedUpdate(
+                      CausalMonotonicityViolation(
+                        contractLedgerEffectiveTime = contractLedgerEffectiveTime,
+                        transactionLedgerEffectiveTime = transactionLedgerEffectiveTime,
+                      )(errorFactories),
+                      transaction.submitterInfo.toCompletionInfo,
+                    )
+                  )
+                )
+              )
+        }
+  }
+
+  private def validateKeyUsages(
+      transaction: Transaction,
+      keyInputs: KeyInputs,
+  )(implicit
+      contextualizedErrorLogger: ContextualizedErrorLogger
+  ): AsyncValidation[Unit] =
+    keyInputs.foldLeft(Future.successful[Validated[Unit]](Right(()))) {
+      case (f, (key, inputState)) =>
+        f.flatMap {
+          case Right(_) =>
+            indexService
+              // TODO SoX: Perform lookup more efficiently and do not use a readers-based lookup
+              .lookupContractKey(transaction.transaction.informees, key)
```

Comment:
As this happens before sequencing will a key that is in-flight be unknown to indexService ?

Reply:
Very good point. This mechanism indeed does reject more than what a sequential committer would. It is a known limitation but with a small impact imho